README
                            Web Page Hit Counter
     __________________________________________________________________

   This is a [1]server side include "counter" function, built in to the
   Apache httpd server.
     __________________________________________________________________

How to set it up

  First, turn your HTML document into a SHTML document.

   The first thing you have to do is adjust the filename extension of your
   web page so that the server recognizes it as containing server-side
   includes. Just change the extension to .shtml. So, for instance, if
   your web page was named "index.html" or "index.htm", your new web page
   should be named "index.shtml". (Other valid extensions for server-side
   include HTML documents on the Why? InterNetworking web server are:
   "-s.html" and "-s.htm".)

  Next, create a file to hold the value of the counter.

   The first thing you have to do is create a file to hold the value of
   the counter. This is an ASCII text file which contains one line on
   which the the counter value is stored. This file must always have the
   extension ".cnt". The file can begin with any name, but we recommend
   that you use the same base name as the .shtml file you intend to use it
   in. So, if your web page was named, "index.shtml," then name the
   counter file, "index.cnt". Put this file on the server, with one line
   containing the number you want the counter to start at.

   When you put the counter file on the server, it must have special
   "access permissions." It must be set read/write by user, group, and
   other so that the HTTP server can read and update it. You can use the
   chmod command to do this from your FTP program or from the Unix shell
   prompt.

  Finally, add the counter to your SHTML document.

   Add the counter directive in your web page where you want the counter
   to appear. This directive looks like a HTML comment. In its simplest
   form, its "<!--#counter file="filename" -->". Filename should be the
   base name of the counter file, without the ".cnt" extension. So, if
   your counter file is named "counter.cnt", the counter directive you
   would use is "<!--#counter file="counter" -->". There you go, you now
   have a counter in your page. Bring up Netscape and try it out!
     __________________________________________________________________

Advanced Options:

   You can do quite a bit more with the counter than just add one to it
   for each hit. You can keep track of page hits silently, adjust the
   appearance of the counter, and even do things like rotating your home
   page through several backgrounds. Here is the list of options you can
   give the counter directive:

   <!--#counter silent="no" width="0" nocommas="no" min="0"
   max="2147000000" rollover="no" increment="1" ith="no" gfx="#"
   file="counter" include="file%i.html" debug="" -->

   Where:
     * silent: Whether or not to show the value of the counter to the
       person accessing the page. yes - Don't show the counter to the
       page's accesser. no - Show the counter to the page's accesser
       (default).
     * width: The minimum width of the displayed counter (padded with 0's
       if necessary).
     * nocommas: Whether or not to insert commas into the number for
       readability. yes - Don't insert commas. no - insert commas
       (default). If width is not 0 then no commas are inserted regardless
       of the setting in nocommas.
     * max: The maximum number to count to. When the counter reaches this
       number, its behavior is controlled by rollover. By default the
       number is near two billion.
     * min: The minimum number to count down to. When the counter reaches
       this number, its behavior is controlled by rollover. By default the
       number is zero.
     * rollover: When you hit the maximum or minimum allowed number, what
       do you do? yes - Flip to the other endpoint. I.e. if at max, change
       to min. If at min, change to max. no - Stop counting. Do not add or
       subtract any further from the counter.
     * increment: The number to add to the counter. Default: 1. You may
       also use the keyword "random" which means that the next value of
       the counter should be a random number between min and max.
     * ith: Whether or not to append st, nd, rd, or th as appropriate to
       the end of the counter (i.e. 1^st, 2^nd, 3^rd, 4^th, and so on.)
       yes - Add the extension. no - Don't add the extension (default).
     * gfx: Use graphics (i.e. GIF files) to display the counter number
       instead of using a text counter. Put the html markup to include the
       graphics where "#" is the digit to replace. For example, if your
       counter were at 1,234 and you used the following line: <!--#counter
       ith="yes" gfx="<IMG SRC=\"gifs/#.gif\">" file="counter" --> then it
       would be replaced by the html markup: <IMG SRC="gifs/1.gif"><IMG
       SRC="gifs/,.gif"><IMG SRC="gifs/2.gif"><IMG SRC="gifs/3.gif"><IMG
       SRC="gifs/4.gif">
     * file: The name of the file (minus the .cnt extension) to store the
       counter in. (For example, if your filename is "home.cnt" then you
       should use file="home".) Warning: the "file" directive must come
       after everything except include and debug.
     * include: The "base" filename (where %i is the counter) of a file to
       include in to the document based on the value of the counter. (For
       example, if you want to include pageX.html where X is the value of
       the counter, use include="page%i.html".)
     * debug, if it exists, means print out the internal variables at the
       point in the command to assist in figuring out why the counter
       isn't working. This is provided for expert users... If you don't
       understand what its saying, don't use it.

   Oh yeah. By the way, you are the person to read this page.

   Enjoy!
            _____________________________________________________

Suggested Uses:

  Basic hit counter

   Add this to your web document:
   Welcome, you are the <strong><!--#counter ith="yes" file="counter"
   --></strong> person to access this web page.

   And create the counter.cnt file. It looks like:
   Welcome, you are the 1,234,567^th person to access this web page.
            _____________________________________________________

Revision History

  v2.4

     * Updated to Apache 2.4

  vA1.1.1-2

     * Added yes/no and 1/0 as valid in addition to true/false

  vA1.1.1-1

     * Added "gfx" option.
     * Upgraded to function with Apache 1.1.1

  v1.5.1-1

     * Added "random" setting to increment option
     * Added "yes/no" options for all tags. Previously the existance of
       the tag with nothing in the quotes meant do the opposite of the
       default. This action is preserved for backward compatibility.
     * Upgraded to function with NCSA httpd 1.5.1

  v1.5a-1

     * Added "ith" option.
     * Upgraded to function with NCSA httpd 1.5a
     __________________________________________________________________



apache-include-counter for Debian
-----------------

build with dpkg-buildpackage
<possible notes regarding this package - if none, delete this file>

cd c
cp filters/mod_include.h filters/mod_include.c .
cp mod_include.c mod_include_counter.c
cp mod_include.h mod_include_counter.h
patch -p0 < mod_include_counter.h.diff
patch -p0 < mod_include_counter.c.diff

 -- William Herrin <herrin@dirtside.com>  Mon, 07 Jan 2013 11:15:38 -0500


debian/changelog
apache-include-counter (2.4.56) unstable; urgency=medium

  * Update for Apache 2.4.56 

 -- William Herrin <herrin@dirtside.com>  Sat, 25 Mar 2023 12:06:15 -0700

apache-include-counter (1.1+1) unstable; urgency=medium

  * update for Apache 2.4

 -- William Herrin <herrin@cat.dirtside.com>  Fri, 03 Jun 2016 11:22:21 -0400

apache-include-counter (1.0-1) unstable; urgency=low

  * apache server side includes w/ counter function
  * import from apache 1.3.1

 -- William Herrin <herrin@dirtside.com>  Mon, 15 Jul 2013 20:19:41 -0400



c/strlwr.c
/* strlwr.c
 *
 * Written by William D. Herrin, herrin@dirtside.com
 * Released to the public domain in 2004.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "lowercase.h"

char *staticstrlwr (const char *ext)
/* Changes ext to lower case and trims at 8 characters. */
{
  static char extlc[11], *p;
  static int i;

  for (i=0,p=extlc; (i<10)&&(*ext!=0) ; i++,ext++,p++)
        *p = lowercasemap[ (int) ((unsigned char) *ext) ];
  *p = 0;
  extlc[8]=0;
  return extlc;
}

int main (int argc, char **argv)
{
  char *s;
  int max, i, cur;

  s = (char*) malloc (sizeof(char)*1000);
  max=1000;
  if (argc>1) {
    cur=0;
    for (i=1; i<argc; i++) {
      if ((strlen(argv[i])+cur)>max) {
        s = (char*) realloc (s,sizeof(char) * ((max+strlen(argv[i]))*2));
      }
      s[cur]=' ';
      strcpy (s+cur+1,argv[i]);
      cur = strlen(s);
    }
    strlwr (s);
    printf ("%s\n",s+1);
  } else {
    while (!feof(stdin)) {
      s[0]=0;
      fgets (s,max,stdin);
      if (s[0]==0) continue;
      strlwr (s);
      printf ("%s",s);
    }
  }
  free (s);
  return 0;
}



c/lowercase.c
/* lowercase.c - functions for converting strings to lower case and making
 * case-insensitive comparisions
 *
 * Written by William D. Herrin, herrin@dirtside.com
 * Released to the public domain in 2004.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>

#include "lowercase.h"

const char *lowercasemap =
      "\000\001\002\003\004\005\006\007\010\011\012\013\014\015\016\017"
      "\020\021\022\023\024\025\026\027\030\031\032\033\034\035\036\037"
      " !\"#$%&'()*+,-./0123456789:;<=>?"
      "@abcdefghijklmnopqrstuvwxyz[\\]^_"
      "`abcdefghijklmnopqrstuvwxyz{|}~\177"
      "cueaaaaceeeiiiaaeaaooouu_ou\233ey\236f"
      "aiounn\246\247?\251\252\253\254\255\256\257"
      "\260\263\262\263\264\265\266\267\270\271\272\273\274\275\276\277"
      "\340\343\342\343\344\345\346\347\350\351\354\353\354\355\356\357"
      "\360\363\362\363\364\365\366\367\370\371\374\373\374\375\376\377"
      "\340\343\342\343\344\345\346\347\350\351\354\353\354\355\356\357"
      "\360\363\362\363\364\365\366\367\370\371\374\373\374\375\376\377";

/* const char *lowercasemap =
      "\000\001\002\003\004\005\006\007\010\011\012\013\014\015\016\017"
      "\020\021\022\023\024\025\026\027\030\031\032\033\034\035\036\037"
      " !\"#$%&'()*+,-./0123456789:;<=>?"
      "@abcdefghijklmnopqrstuvwxyz[\\]^_"
      "`abcdefghijklmnopqrstuvwxyz{|}~\177"
      "\200\201\202\203\204\205\206\207\210\211\212\213\214\215\216\217"
      "\220\223\222\223\224\225\226\227\230\231\232\233\234\235\236\237"
      "\240\243\242\243\244\245\246\247\250\251\252\253\254\255\256\257"
      "\260\263\262\263\264\265\266\267\270\271\272\273\274\275\276\277"
      "\340\343\342\343\344\345\346\347\350\351\354\353\354\355\356\357"
      "\360\363\362\363\364\365\366\367\370\371\374\373\374\375\376\377"
      "\340\343\342\343\344\345\346\347\350\351\354\353\354\355\356\357"
      "\360\363\362\363\364\365\366\367\370\371\374\373\374\375\376\377"; */

char *strlwr (char *s)
/* Changes s to lower case */
{
  char *p;

  for (p=s; (*p!=0); p++) *p = lowercasemap[ (int) ((unsigned char) *p) ];
  return s;
}

int strcmpi (const char *s1, const char *s2)
{
  char ch1, ch2;

  if ((s1==NULL)&&(s2==NULL)) return 0;
  if (s1==NULL) return -1;
  if (s2==NULL) return 1;
  while (1) {
    ch1=lowercasemap[ (int) ((unsigned char) *s1) ];
    ch2=lowercasemap[ (int) ((unsigned char) *s2) ];
    if (ch1 < ch2) return -1;
    if (ch1 > ch2) return 1;
    if (ch1==0) return 0;
    s1++;
    s2++;
  }
}

int strnncmpi (const char *s1, const char *s2, int max)
  /* Compare s1 and s2 in as many as max characters. If they are the same for
   * that duration (or until eol) then return so. Else return less or greater.
   * Comparison is case insensitive.
   */
{
  int i;
  char c1, c2;

  for (i=0; i<max; i++, s1++, s2++) {
    c1=lowercasemap[ (int) ((unsigned char) *s1) ];
    c2=lowercasemap[ (int) ((unsigned char) *s2) ];
    if (c1<c2) return -1;
    if (c1>c2) return 1;
    if (c1==0) return 0;
  }
  return 0;
}

int strnncmp (const char *s1, const char *s2, int max)
  /* Compare s1 and s2 in as many as max characters. If they are the same for
   * that duration (or until eol) then return so. Else return less or greater.
   */
{
  int i;

  for (i=0; i<max; i++) {
    if (s1[i]<s2[i]) return -1;
    if (s1[i]>s2[i]) return 1;
    if (s1[i]==0) return 0;
  }
  return 0;
}



c/mod_include_counter.h
/* Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * @file mod_include_counter.h
 * @brief Server Side Include Filter Extension Module for Apache w/ Counter
 *
 * @defgroup MOD_INCLUDE_COUNTER mod_include_counter
 * @ingroup APACHE_MODS
 * @{
 */

#ifndef _MOD_INCLUDE_COUNTER_H
#define _MOD_INCLUDE_COUNTER_H 1

#include "apr_pools.h"
#include "apr_optional.h"
#include "util_filter.h"

/*
 * Constants used for ap_ssic_get_tag_and_value's decode parameter
 */
#define SSI_VALUE_DECODED 1
#define SSI_VALUE_RAW     0

/*
 * Constants used for ap_ssic_parse_string's leave_name parameter
 */
#define SSI_EXPAND_LEAVE_NAME 1
#define SSI_EXPAND_DROP_NAME  0

/*
 * This macro creates a bucket which contains an error message and appends it
 * to the current pass brigade
 */
#define SSI_CREATE_ERROR_BUCKET(ctx, f, bb) APR_BRIGADE_INSERT_TAIL((bb), \
    apr_bucket_pool_create(apr_pstrdup((ctx)->pool, (ctx)->error_str),    \
                           strlen((ctx)->error_str), (ctx)->pool,         \
                           (f)->c->bucket_alloc))

/*
 * These constants are used to set or clear flag bits.
 */
#define SSIC_FLAG_PRINTING         (1<<0)  /* Printing conditional lines. */
#define SSIC_FLAG_COND_TRUE        (1<<1)  /* Conditional eval'd to true. */
#define SSIC_FLAG_SIZE_IN_BYTES    (1<<2)  /* Sizes displayed in bytes.   */
#define SSIC_FLAG_NO_EXEC          (1<<3)  /* No Exec in current context. */

#define SSIC_FLAG_SIZE_ABBREV      (~(SSIC_FLAG_SIZE_IN_BYTES))
#define SSIC_FLAG_CLEAR_PRINT_COND (~((SSIC_FLAG_PRINTING) | \
                                     (SSIC_FLAG_COND_TRUE)))
#define SSIC_FLAG_CLEAR_PRINTING   (~(SSIC_FLAG_PRINTING))

/*
 * The public SSI context structure
 */
typedef struct {
    /* permanent pool, use this for creating bucket data */
    apr_pool_t  *pool;

    /* temp pool; will be cleared after the execution of every directive */
    apr_pool_t  *dpool;

    /* See the SSIC_FLAG_XXXXX definitions. */
    int          flags;

    /* nesting of *invisible* ifs */
    int          if_nesting_level;

    /* if true, the current buffer will be passed down the filter chain before
     * continuing with next input bucket and the variable will be reset to
     * false.
     */
    int          flush_now;

    /* argument counter (of the current directive) */
    unsigned     argc;

    /* currently configured error string */
    const char  *error_str;

    /* currently configured time format */
    const char  *time_str;

    /* the current request */
    request_rec  *r;

    /* pointer to internal (non-public) data, don't touch */
    struct ssi_internal_ctx *intern;

} include_counter_ctx_t;

typedef apr_status_t (include_counter_handler_fn_t)(include_counter_ctx_t *, ap_filter_t *,
                                            apr_bucket_brigade *);

APR_DECLARE_OPTIONAL_FN(void, ap_ssic_get_tag_and_value,
                        (include_counter_ctx_t *ctx, char **tag, char **tag_val,
                         int dodecode));

APR_DECLARE_OPTIONAL_FN(char*, ap_ssic_parse_string,
                        (include_counter_ctx_t *ctx, const char *in, char *out,
                         apr_size_t length, int leave_name));

APR_DECLARE_OPTIONAL_FN(void, ap_register_include_counter_handler,
                        (char *tag, include_counter_handler_fn_t *func));

#endif /* MOD_INCLUDE_COUNTER */
/** @} */


c/counter.c
/* counter.c - standalone functions supporting the counter inside
 * an apache server side include
 */

#include <stdio.h> /* FILE */
#include <stdlib.h> /* NULL */
#include <string.h> /* strcpy */
#include <sys/types.h> /* off64_t in apr.h */
#include <time.h> /* time() */
#include <sys/stat.h> /* open() */
#include <fcntl.h> /* open() */
#include <unistd.h> /* write() */

#include "lowercase.h"
#include "httpd.h" /* MAX_STRING_LEN */
#include "http_log.h" /* APLOG_MARK */
#include "http_request.h" /* ap_sub_req_lookup_file */
#include "http_config.h" /* ap_set_module_config */
#include "mod_include_counter.h"
#include "apr_strings.h" /* apr_pstrdup */

#include "counter.h"

/* -------------------- Support Functions for Counter --------------------- */


extern module AP_MODULE_DECLARE_DATA include_counter_module;

char *strcpyupto (char *dest, const char *source, int max)
{
  int i;
  if (source==NULL) { 
    dest[0]=0;
    return dest;
  }
  for (i=0;(source[i]!=0)&&(i<max);i++) dest[i]=source[i];
  dest[i]=0;
  return dest;
}

void splitagent (char *agent, char *name, char *vers)
/* Take the user-agent string, such as:
 *   Mozilla/1.2N (Windows; I; 32bit)
 * and split the part before the / (Mozilla) into name, and the first 5 digits
 * (padded by zeros) after the / (12320) into vers.
 */
{
  int i,j;

  for (i=0;(agent[i]!=0)&&(agent[i]!='/')&&(i<250);i++) 
    name[i]=agent[i];
  name[i]=0;
  strcpy (vers,"00000");
  for (j=0;(agent[i]!=0)&&(j<5);i++) 
    if ((agent[i]>='0')&&(agent[i]<='9')) vers[j++]=agent[i];
  return;
}

int checkagent (char *agent, char *name, char *vers, char *add) 
/* Check the agent description against the actual agent name, and if it matche
s
 * then set add to the appropriate string to add to the string and return 1,
 * else return 0.
 * agent="MatchThisUserAgent/InsertThisIntoFilename/NumberOfVersionDigits"
 */
{
  int i,j,k;

  for (i=0;(agent[i]!=0)&&(agent[i]!='/')&&(i<250);i++) ;
  if ((agent[i]!='/')||(strnncmpi (agent,name,i)!=0)) return 0;
  for (i=i+1,j=0;(j<20)&&(agent[i]!=0)&&(agent[i]!='/');j++,i++)
    add[j] = agent[i];
  add[j]=0;
  if (agent[i]!='/') return 1;
  k=(int) strtol (agent+i+1,NULL,10);
  if (k>5) k=5;
  for (i=0;i<k;i++) add[i+j]=vers[i];
  add[i+j]=0;
  return 1;
}

/* The Counter modification is Copyright 1996-2005 William Herrin.
 * Its offered under the Apache License, Version 2.0
 * Visit my home page at http://bill.herrin.us/ or email me at
 * herrin@dirtside.com.
 *
 * This is counter patch revision 1.3.33-2
 */



/* #define LOGSSI        1 */
#ifdef LOGSSI
int ap_log_transaction (request_rec *r);
/* Aparantly I'm really not supposed to be calling log_transaction from
 * here. */

char *fixloggedssi (char *buffer, const char *lastreq, const char *virtual_url)
{
  char oldreq[MAX_STRING_LEN];
  char *p;

  if ((lastreq==NULL)||(virtual_url[0]=='/')) {
    sprintf (buffer, "SSI %s INCLUDE/0.0", virtual_url);
    return buffer;
  }
  strcpy (oldreq,lastreq);
  for (p=oldreq; (*p!='?') && (*p!=' ') && (*p!=0); p++);
  *p=0;
  for (; (*p!='/') && (p>oldreq); p--) ;
  if (*p=='/') strcpy (p+1,virtual_url);
  sprintf (buffer, "SSI %s INCLUDE/0.0", oldreq);
  return buffer;
}
#endif

int checkyesno (char *s)
/* Check if S contains yes|true|1 versus no|false|0 and return 1 or 0.
 * Return -1 if the value is invalid. */
{
  if (s==NULL) return -1;
  switch (s[0]) {
    case 'y':
    case 't':
    case 'Y':
    case 'T':
    case '1':
      return 1;
    case 'n':
    case 'f':
    case 'N':
    case 'F':
    case '0':
      return 0;
    default:
      return -1;

  }
}

void ap_ssic_get_tag_and_value(include_counter_ctx_t *ctx, char **tag,
                                     char **tag_val, int dodecode);
    
static void counter_error (
  include_counter_ctx_t *ctx, /* for gettag */
  ap_filter_t *f,
  apr_bucket_brigade *bb,
  const char *fmt,
  ...
) {
  char output[MAX_STRING_LEN], *p;
  size_t len;
  request_rec *r = f->r;
  va_list argptr;

  va_start (argptr,fmt);
  vsnprintf (output,MAX_STRING_LEN,fmt,argptr);
  va_end (argptr);
  for (p=output; *p; p++) if (*p=='%') *p='_';
  len = (size_t) (p-output);

  ap_log_rerror(APLOG_MARK, APLOG_NOERRNO|APLOG_ERR, 0, r,"%s",output);
  APR_BRIGADE_INSERT_TAIL(bb, apr_bucket_pool_create(
                                apr_pmemdup(ctx->pool, output,len),
                                len, ctx->pool, f->c->bucket_alloc));
  return;
}

int handle_counter(
  include_counter_ctx_t *ctx, /* for gettag */
  ap_filter_t *f,
  apr_bucket_brigade *bb
) {
    char *tag, *tag_val;
    char fn[MAX_STRING_LEN+4],*p, gfx[MAX_STRING_LEN];
    char s[MAX_STRING_LEN+6], sr[MAX_STRING_LEN+6];
    int count=0,       /* The counter itself */
      wide=0,        /* Pad to this many zeros */
      silent=0,      /* 0 = tell number, 1 = only increment, don't display */
      increment=1,   /* How much to increment by */
      max=2147000000L,  /* How high can the number go? */
      min=0,         /* How low can the number go? */
        commas=1,      /* Use commas by default */
      xth=0,         /* Do not include th after number by default */
        random=0,      /* Instead of incrementing, jump to a random amount */
      dontincr=0,    /* Don't increment because file is not right */
      rollover=0,    /* Restart at min if above max or max if below min? */
      filedes=0;     /* File descriptor for write */
    int i,j; /* Generic Counters */
    char *ith;
    request_rec *r = f->r;
    int counted=0; /* did I actually load a counter? */


    fn[0]=0;s[0]=0; p=s;
    strcpy (gfx,"#");

    while(1) {
        ap_ssic_get_tag_and_value(ctx,&tag,&tag_val,0);
        if((!tag || strcmp(tag,"done")==0)) {
          /* close of <!--#counter tag */
           if (!counted) {
            counter_error (ctx,f,bb,
                "[Counter: no valid file= in #counter]");
          }
          return APR_SUCCESS;
        }
        if(!tag_val) {
          counter_error (ctx,f,bb,
                "[Counter: unknown error in #counter]");
          return APR_SUCCESS;
        }
        if (strcmp (tag,"silent") ==0 ) {
          if (tag_val[0]==0) silent = 1-silent;
          if ((silent=checkyesno(tag_val))<0) {
            silent=0;
            counter_error (ctx,f,bb,
                "[Counter: silent=\"%s\" is invalid."
                " It should be either true or false in %s]",
                            tag_val, r->filename);
          }
          continue;
        }
        if (strcmp (tag,"increment") ==0 ) {
          if ((tag_val[0]=='r')||(tag_val[0]=='R')) random=1;
          else {
            increment = (int) strtol (tag_val,NULL,10);
            random=0;
          }
          continue;
        }
        if (strcmp (tag,"width") ==0 )
          wide = (int) strtol (tag_val,NULL,10);
        else if (strcmp (tag,"max") ==0 )
          max = (int) strtol (tag_val,NULL,10);
        else if (strcmp (tag,"min") ==0 )
          min = (int) strtol (tag_val,NULL,10);
        else if (strcmp (tag,"rollover") ==0 ) {
          if (tag_val[0]==0) rollover = 1-rollover;
          else if ((rollover=checkyesno(tag_val))<0) {
            rollover=0;
            counter_error (ctx,f,bb,
              "[Counter: rollover=\"%s\" is invalid."
              " It should be either true or false in %s]",
              tag_val,r->filename);
          }
        }
        else if (strcmp (tag,"nocommas") ==0 ) {
          if (tag_val[0]==0) commas = 1-commas;
          else if ((commas=checkyesno(tag_val))<0) {
            commas=0;
            counter_error (ctx,f,bb,
                "[Counter: nocommas=\"%s\" is invalid."
                " It should be either true or false in %s]", 
                tag_val,r->filename);
          }
        }
       else if (strcmp (tag,"ith") ==0 ) {
          if (tag_val[0]==0) xth = 1-xth;
          else if ((xth=checkyesno(tag_val))<0) {
            xth=0;
            counter_error (ctx,f,bb,
                "[Counter: xth=\"%s\" is invalid."
                " It should be either true or false in %s]",
                 tag_val,r->filename);
          }
        }
        else if (strcmp (tag,"gfx") ==0 ) {
          strncpy (gfx,tag_val,MAX_STRING_LEN);
          gfx[MAX_STRING_LEN-10]=0; /* leaving room for the counter value */
        }
        else if (strcmp (tag,"debug") ==0 ) {
          counter_error (ctx,f,bb,
              "[Counter Debug: Commas=%s, Silent=%s, Max=%i, Min=%i, "
              "RollOver=%s, Width=%i, IncrementBy=%i, Counter=%i, "
              "Filename=\"%s\"]",
              (commas==0)?"No":"Yes",(silent==0)?"No":"Yes",max,min,
                (rollover==0)?"No":"Yes",wide,increment,count,fn);
        }
        else if (strcmp (tag,"include") ==0 )
        { /* Include a file based on the counter */
          char url[MAX_STRING_LEN];
          char *newpath;
          apr_status_t rv;
          char *error_fmt = NULL;
          request_rec *rr = NULL;

          strncpy (s,tag_val,MAX_STRING_LEN);
          s[MAX_STRING_LEN-10]=0; /* leaving room for the counter value */
          for (i=0,j=0;s[i]!=0;i++)
            if (s[i]=='%') {
              if (j!=0) s[i]='-';
              else {
                if (s[i+1]!=0) s[i+1]='i';
                else s[i]='-';
                j=1;
              }
            }
          if (j==0) strcat (s,"%i");
          sprintf (url,s,count);

          /* be safe; only files in this directory or below allowed */
          rv = apr_filepath_merge(&newpath, NULL, url,
                                    APR_FILEPATH_SECUREROOTTEST |
                                    APR_FILEPATH_NOTABSOLUTE, ctx->dpool);

          if (rv != APR_SUCCESS) {
                error_fmt = "unable to include file \"%s\" in parsed file %s";
          } else {
                rr = ap_sub_req_lookup_file(newpath, r, f->next);
          }
          if (!error_fmt && rr->status != HTTP_OK) {
            error_fmt = "unable to include \"%s\" in parsed file %s";
          }
          if (!error_fmt && (ctx->flags & SSIC_FLAG_NO_EXEC) &&
            rr->content_type && strncmp(rr->content_type, "text/", 5)) {

            error_fmt = "unable to include potential exec \"%s\" in parsed "
                        "file %s";
          }
          if (rr) {
            ap_set_module_config(rr->request_config, &include_counter_module,r);
          }
        if (!error_fmt && ap_run_sub_req(rr)) {
            error_fmt = "unable to include \"%s\" in parsed file %s";
        }
        if (error_fmt) {
            counter_error (ctx,f,bb,
                 error_fmt, tag_val, r->filename);
        }
        /* Do *not* destroy the subrequest here; it may have allocated
         * variables in this r->subprocess_env in the subrequest's
         * r->pool, so that pool must survive as long as this request.
         * Yes, this is a memory leak. */

#ifdef LOGSSI
          strcpy (s,"SSI ");
          if (url[0]!='/') {
            strncat (s,url,MAX_STRING_LEN-4);
            for (p = s+strlen(s)-1;(p>s+4)&&(*p!='/');p--) *p=0;
          }
          strncat (s,url,MAX_STRING_LEN-strlen(s));
          s[MAX_STRING_LEN-1]=0;
          if (strlen(s)<MAX_STRING_LEN-15) strcat (s," INCLUDE/1.5.1");
          eog_generic(reqInfo,s,0,reqInfo->url);
#endif /* LOGSSI */
        }
        else if (strcmp (tag,"file") ==0 )
        { /* Increment and maybe display the counter */
          FILE *cf;
          char *fnp,*fnslash;

          strncpy (fn,r->filename,MAX_STRING_LEN);
          for (fnp=fnslash=fn; *fnp; fnp++) {
            if (*fnp=='/') fnslash=fnp;
          }
          snprintf (fnslash+1,MAX_STRING_LEN-((int)(fnslash-fn))-10,
            "%s.cnt",tag_val);
          sprintf (s,"%s/",fn);
          if (strstr(s, "/../") != NULL) {
            counter_error (ctx,f,bb,
              "[Counter: %s is not in or beneath the current directory "
              "in %s]",fn, r->filename);
            continue;
          }
          cf = fopen (fn,"rt");
          if (cf==NULL)
          {
            counter_error (ctx,f,bb,
                "[Counter: %s not found in %s]", fn,r->filename);
            continue;
          }
          s[0]=0;
          /*fscanf (f,"%i",&count);*/
          if( fgets (s,MAX_STRING_LEN,cf) );
          if ((s[0]>='0')&&(s[0]<='9')) count = (int) strtol (s,NULL,10);
          else {
            counter_error (ctx,f,bb,
                "[Counter: %s does not contain a number in %s]",
                fn,r->filename);
            dontincr = 1;
            continue;
          }
          fclose (cf);
          counted = 1; /* 1 means true */
          if (!random)
            count+=increment;
          else {
            srand((int) time(NULL));
            count = rand();
            count %= max+1-min;
            count += min;
          }
          if (count>max)
          {
            if (rollover==0) count=max;
            else count-=(max-min+1);
          }
          if (count<min)
          {
            if (rollover==0) count=min;
            else count+=(max-min+1);
          }
          if (!silent)
          {
            if (wide!=0) /* minimum width, use no commas */
            {
              sprintf (s,"%%.%ii",wide);
              sprintf (sr,s,count);
            /*rprintf(reqInfo,s,count);*/
            }
            else if (commas==0) /* No commas */
              /*rprintf(reqInfo,"%i",count);*/
              sprintf(sr,"%i",count);
            else /* Use commas */
            {
              i=count;
              j=0;
              if (i<0) {
                i=0-i;
                j=1;
              }
              s[20]=0;
              p=s+17;
              sprintf (p,"%.3i",i%1000);
              i/=1000;
              for (;i>0;i/=1000)
              {
                sprintf (p-4,"%.3i",i%1000);
                p--;
                *p=',';
                p-=3;
              }
              while (*p=='0') p++;
              if (*p==0) {
                 p--;
                 *p='0';
              }
              if (j) {
                p--;
                *p='-';
              }
              sprintf(sr,"%s",p);
              /*rprintf(reqInfo,"%s",p);*/
            }
            if (xth) {
              ith="th";
              i=count;
              if (i<0) i=0-i;
              if (((i%10)==1)&&(((i%100)/10)!=1)) ith="st";
              if (((i%10)==2)&&(((i%100)/10)!=1)) ith="nd";
              if (((i%10)==3)&&(((i%100)/10)!=1)) ith="rd";
              /*rprintf(reqInfo,"<SUP>%s</SUP>",ith);*/
              if (strcmp (gfx,"#")==0) 
                sprintf(sr+strlen(sr),"<SUP>%s</SUP>",ith);
              else sprintf(sr+strlen(sr),"%s",ith);
            }
            if (strcmp (gfx,"#")==0)
              APR_BRIGADE_INSERT_TAIL(bb, apr_bucket_pool_create(
                                apr_pmemdup(ctx->pool, sr,strlen(sr)),
                                strlen(sr), ctx->pool, f->c->bucket_alloc));
            else {
              for (p=sr; *p!=0; p++) {
                strcpy (s,gfx); 
                for (i=0; s[i]!=0; i++) if (s[i]=='#') s[i]=*p;
                APR_BRIGADE_INSERT_TAIL(bb, apr_bucket_pool_create(
                                apr_pmemdup(ctx->pool, s,strlen(s)),
                                strlen(s), ctx->pool, f->c->bucket_alloc));
              }
            }
          }
          if (!dontincr) {
            /* Note: use open/write instead of fopen/fprintf so that we
             * we don't clobber the counter if the disk fills up. */
            filedes = open (fn,O_RDWR);
            if (filedes<0)
            {
              counter_error (ctx,f,bb,
                "[Counter: Cannot write to %s; Cannot increment in %s]",
                fn, r->filename);
              continue;
            }
            sprintf (s,"%i\n",count); /* MAX_STRING > possible int size */
            if (write (filedes,(void*) s,(strlen(s)*sizeof(char) ))) ;
            close (filedes);
          }
          /*    ap_log_rerror(APLOG_MARK, APLOG_NOERRNO|APLOG_ERR, 0, r,
                "[Counter: checkpoint in %s]",r->filename); */
          continue;
          /* return APR_SUCCESS; */
        }
        else {
          counter_error (ctx,f,bb,
            "[Counter: Unknown tag %s=%s in %s]",
                 tag,tag_val,r->filename);
          return APR_SUCCESS;
        }
    }

    return APR_SUCCESS;
}

